### PR TITLE
Allow buck http cache verify http header

### PR DESCRIPTION
**Pull request by zayhero-zz**
Right now buck http cache didn't do any auth check against http header. To ensure we are safe from others uploading buggy artifacts, this PR is adding the logic to verify the header.

This PR includes:
1. config change to introduce `authorizationConfig`, so that user can send the `authorizedTokens` over
2. introduce `HttpHeaderAuthFilter ` to do http header authentication, what it does is extract the `Authorization` from Http header, and verify if that contains in `authorizedTokens` from config
3. change `BuckCacheApplication` to register the `HttpHeaderAuthFilter` so that it could be used
4. change `BuckCacheResource` to execute the authentication for **PUT** request
5. update existing and new unit test to verify the change

To use that, it require:
1. add authorized token into `http_write_headers` in `.buckconfig`, e.g.
```
[cache]
  mode = http
  http_mode = readwrite
  http_url = http://localhost:6457
  http_write_headers = Authorization: ZAYHERO
```
2. add the token (i.e. `ZAYHERO`) into `standalone.yml`, e.g.
```
authorizationConfig:
  authorizedTokens:
    - ZAYHERO
```

@dhaval2025 @kageiit Please take a look

### CODE REVIEW

File: cache/src/main/java/com/uber/buckcache/auth/HttpHeaderAuthFilter.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import javax.annotation.Priority;
+import javax.ws.rs.NotAuthorizedException;
+import javax.ws.rs.Priorities;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.core.HttpHeaders;
+
+@Priority(Priorities.AUTHENTICATION)
+public class HttpHeaderAuthFilter extends AuthFilter {
+
+  protected List<String> authorizedToken;
```
Comment (kageiit) [inline cache/src/main/java/com/uber/buckcache/auth/HttpHeaderAuthFilter.java]: `authorizedTokens`

File: cache/src/main/java/com/uber/buckcache/auth/HttpHeaderAuthFilter.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  @Override
+  public void filter(final ContainerRequestContext requestContext) throws IOException {
+    if (authorizedToken == null) {
+      return ;
+    }
+
+    String token = requestContext.getHeaders().getFirst(HttpHeaders.AUTHORIZATION);
+
+    if (!authorizedToken.contains(token)) {
```
Comment (kageiit) [inline cache/src/main/java/com/uber/buckcache/auth/HttpHeaderAuthFilter.java]: check if `token` is null first

File: cache/src/main/java/com/uber/buckcache/resources/buckcache/BuckCacheResource.java [hunk 3]
```diff
@@ -111,7 +113,8 @@ public Response getCacheArtifact(@PathParam("key") String key) throws Exception
   @Path("key")
   @Consumes(MediaType.APPLICATION_OCTET_STREAM)
   public Response addArtifactToCache(PutCacheEntry putCacheEntry,
-                                     @HeaderParam(X_CACHE_EXPIRY_SECONDS) String cacheExpirySeconds) throws Exception {
+                                     @HeaderParam(X_CACHE_EXPIRY_SECONDS) String cacheExpirySeconds,
+                                     @Auth @HeaderParam(HttpHeaders.AUTHORIZATION) String auth) throws Exception {
```
Comment (kageiit) [inline cache/src/main/java/com/uber/buckcache/resources/buckcache/BuckCacheResource.java]: where is the `auth` used here?

File: cache/src/test/java/com/uber/buckcache/auth/HttpHeaderAuthFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  private static MultivaluedMap mockHeaders;
+
+  @BeforeClass
+  public static void setup() throws Exception {
+    mockRequestContext = mock(ContainerRequestContext.class);
+    mockHeaders = new MultivaluedHashMap<String, String>();
+    mockHeaders.add(HttpHeaders.AUTHORIZATION, "testToken");
+    when(mockRequestContext.getHeaders()).thenReturn(mockHeaders);
+  }
+
```
Comment (kageiit) [inline cache/src/test/java/com/uber/buckcache/auth/HttpHeaderAuthFilterTest.java:33]: Can you add a test for failure without a token as well?

**Review by kageiit — CHANGES_REQUESTED** [review]

**Comment by kageiit** [comment]
Can we change `Authorization` to be called `Authentication` instead, in the header?

File: cache/src/main/java/com/uber/buckcache/auth/HttpHeaderAuthFilter.java (hunk 1 above)
Comment (zayhero-zz) [reply cache/src/main/java/com/uber/buckcache/auth/HttpHeaderAuthFilter.java]: updated

File: cache/src/main/java/com/uber/buckcache/auth/HttpHeaderAuthFilter.java (hunk 2 above)
Comment (zayhero-zz) [reply cache/src/main/java/com/uber/buckcache/auth/HttpHeaderAuthFilter.java]: updated, add `token == null` checking

File: cache/src/main/java/com/uber/buckcache/resources/buckcache/BuckCacheResource.java (hunk 3 above)
Comment (zayhero-zz) [reply cache/src/main/java/com/uber/buckcache/resources/buckcache/BuckCacheResource.java]: its used by `HttpHeaderAuthFilter`, but I change the code to call `@PermitAll` to activate the authentication

File: cache/src/test/java/com/uber/buckcache/auth/HttpHeaderAuthFilterTest.java (hunk 4 above)
Comment (zayhero-zz) [reply cache/src/test/java/com/uber/buckcache/auth/HttpHeaderAuthFilterTest.java:33]: add `testFailWithoutAuthorizedToken` to test this

**Review by zayhero-zz — COMMENTED** [review]
comment addressed, @kageiit please take a look :)

File: cache/src/dist/config/standalone.yml [hunk 5]
```diff
@@ -27,6 +27,9 @@ storeProviderConfig:
     atomicSequencereserveSize: 10000
     offHeapStorageSize: 40g # 40GB
 
+authenticationConfig:
+  authenticatedTokens:
```
Comment (kageiit) [inline cache/src/dist/config/standalone.yml]: just call these `tokens`

File: cache/src/main/java/com/uber/buckcache/AuthenticationConfig.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package com.uber.buckcache;
+
+import java.util.List;
+import javax.annotation.Nonnull;
+
+public class AuthenticationConfig {
+  @Nonnull
+  private List<String> authenticatedTokens;
```
Comment (kageiit) [inline cache/src/main/java/com/uber/buckcache/AuthenticationConfig.java]: `tokens`

**Comment by kageiit** [comment]
Minor renames, but looks good otherwise. Please make the changes. I will approve and merge after the renames

**Review by kageiit — CHANGES_REQUESTED** [review]
minor changes

**Comment by zayhero-zz** [comment]
@kageiit make further renames to s/authenticatedTokens/tokens , please take a look

**Review by kageiit — APPROVED** [review]

**Comment by kageiit** [comment]
LGTM, but would like @dhaval2025 to make a sanity check

**Comment by fkorotkov** [comment]
@kageiit maybe check both `Authorization` and `Authentication` headers? Buck documentation uses `Authorization` in their examples: https://buckbuild.com/concept/buckconfig.html#cache.http_write_headers

Update: my bad. I misunderstood comments but then I checked the diff. `Authorization` header is checked but `Authentication` used in naming.

**Comment by dhaval2025** [comment]
looks good. Merging it in ..

**Pull request merged**